### PR TITLE
[ENH] Add SetChannelsMixin to Info

### PR DESCRIPTION
**Pull request by alexrockhill**
If we could add the `SetChannelsMixin` to `Info` that would be really nice. I find myself wanting to change the channel type of `Info` objects when using `Array` classes all the time. Like the other mixin in meas_info, it only operates on the `Info` so I think, in general, when methods don't use the data, it's nice if they could work for `Info` too.

### CODE REVIEW

**Review by agramfort — APPROVED** [review]
LGTM

**Review by hoechenberger — COMMENTED** [review]
Looks good, but could you add a changelog entry, please?

**Review by drammock — COMMENTED** [review]
I was very surprised looking at the diff: I expected just a few added lines and see instead `+444 -419`. This is because your PR title and PR description do not match what you're actually doing, which is:

1. Add a new `ReferenceMixin` class that contains only the `set_eeg_reference` method
2. Remove `set_eeg_reference` from `SetChannelsMixin`
3. Move the `SetChannelsMixin` from `mne/channels/channels` to `mne/io/meas_info`
4. Add the `ReferenceMixin` everywhere that `SetChannelsMixin` is currently used
5. Add `SetChannelsMixin` to `Info`

Once one figures out what's being done it's all fairly logical, but it would save your reviewers a lot of time / effort / head-scratching if you would take the time to explain *what you've actually done* (as I've just done for you) instead of explaining *what you want to be able to do* --- that would have been appropriate if you were first opening an *issue* to discuss your desired changes rather than jumping straight to a PR implementation.

Other than that meta-commentary, I'll add that moving `SetChannelsMixin` to a new file makes sense but IMO is not a great idea, for two reasons: (1) it messes with `git blame` making it harder to delve into the history of that class, and (2) it's harder to review (lots of extra red/green lines, hard to tell if they're just copy/paste or if any changes were snuck in there).

You can potentially mitigate both of those by separating your commits cleanly:

1. move class to other file (and fix relative imports accordingly)
2. add new class w/ set_eeg_reference method, remove method from original class, and add new class to all the objects that need it

...then adding the move-class-to-other-file commit to `.git-blame-ignore-revs`, and requesting a rebase-merge instead of squash-merge. (Structuring the commits this way is important for rebase-merges because it leaves the code in a usable state after each commit.) Are you comfortable enough with git to give that a try?

**Comment by alexrockhill** [comment]
Ok, now the commit history should be clearer with one copy-paste move over and then the subsequent changes and the git-blame will ignore the copy-paste.

**Comment by alexrockhill** [comment]
Okay now looks easier to review and the blame should be ok

File: mne/io/meas_info.py
Comment (drammock) [inline mne/io/meas_info.py:477]: it is exactly the small changes like this that are easy to see when the copy-paste is done in a separate commit, and very hard to see otherwise.  (this looks fine, BTW)

**Review by drammock — COMMENTED** [review]
Thanks for doing that!  Way easier to review now. Remaining things I noticed:

- check / fix the failing CIs (which I haven't checked, not sure if related)
- please add a comment to the new line in `.git-blame-ignore-revs`, to the effect of "copy-paste move between files of the SetChannelsMixin class" (or something equivalent but less wordy :)
- Note that the commits aren't split exactly as requested: the first commit includes only the copy-paste *and not the adjustments to the import statements in other files* so running MNE off that commit would be broken (all class constructors that inherit from SetChannelsMixin would fail). This is IMO not as important as facilitating easy review, but it can still be an issue when running `git bisect` for example: it could cause spurious failures unrelated to whatever other failure one is trying to isolate.

File: .git-blame-ignore-revs
Comment (agramfort) [inline .git-blame-ignore-revs:2]: what is this?

**Review by agramfort — COMMENTED** [review]
just a nit

I had already wrapped my head around the first pass.

I'll let @drammock do the 2nd review.

**Comment by drammock** [comment]
@alexrockhill I think this is the remaining unaddressed bit:

> Note that the commits aren't split exactly as requested: the first commit includes only the copy-paste _and not the adjustments to the import statements in other files_ so running MNE off that commit would be broken (all class constructors that inherit from SetChannelsMixin would fail). This [...] can still be an issue when running `git bisect` [...]

@larsoner just flagging again that this one is destined for a rebase-merge due to a large copy-paste between files, so we need a clean commit sequence before merging.

**Comment by alexrockhill** [comment]
I wasn't sure how to do that with `git`

**Comment by drammock** [comment]
> I wasn't sure how to do that with `git`

ok no worries. FYI, I did it myself with this approach:

1. `git checkout -b info alexrockhill/info`  (you would skip this step)
2. `git reset --hard 6ae1ec4ea`  # the hash of the commit parent to this branch
3. `git rebase upstream/main`  # so we're working off current main, not main of several days ago
4. `git cherry-pick cccb72f5f`  # the big copy-paste commit
5. `git cherry-pick -n 6ae1ec4ea`  # the "fix diff" commit in this PR. `-n` means that cherry-pick doesn't automatically create the commit.
6. `git checkout HEAD .git-blame-ignore-revs`  # hash will change anyway so we'll need to redo this later
7. `git restore --staged mne/io/meas_info.py mne/io/base.py mne/evoked.py mne/epochs.py mne/channels/channels.py` # unstage everything else
8. `git add -Ap`  # add back changes from the "fix diff" commit in "patch mode" --- add in everything needed to keep things working on this commit. This means adjusting just the imports (generally speaking: changing `from .channels.channels` to `from .io.meas_info` in a few places)
9. `git commit -m "adjust imports"` # (temporary, this will get squashed later)
10. `git add -Ap` # now add the changes that create the new `ReferenceMixin` and removes `set_eeg_channels` from the old mixin
11. `git commit -m "move set_eeg_ref to its own mixin"`  # (temporary, will get squashed)
12. `git add -Ap` # now finally add the changes to `SetChannelsMixin` that allow it to work with `Info`, and add it as a parent class to `Info`
13. `git commit -m "add SetChannelsMixin to Info"`

Result is now:

```
b3e5db209 (HEAD -> info) (4 minutes ago) dan@mccloy.info add SetChannelsMixin to Info
2a0c65c32 (6 minutes ago) dan@mccloy.info move set_eeg_reference to its own mixin
d88c39800 (6 minutes ago) dan@mccloy.info adjust all the imports
b497f5d07 (10 minutes ago) dan@mccloy.info move setchannelsmixin to meas_info.py
```

Now all that remains is to squash the bottom 2 commits together, add back the correct hash to `.git-blame-ignore-revs`, and then add that and squash it with the top two commits.

```
git rebase -i b497f5d07^  # mark d88c39800 as "fixup"
# now we have
5d58e9885 (HEAD -> info) (2 seconds ago) dan@mccloy.info add SetChannelsMixin to Info
5117ab219 (3 seconds ago) dan@mccloy.info move set_eeg_reference to its own mixin
c2fbb9f1a (3 seconds ago) dan@mccloy.info move setchannelsmixin to meas_info
```
so now we add c2fbb9f1a to the blame ignore file and do another `rebase -i` with a couple of "fixups" and one "reword", to get:
```
ea19a825d (HEAD -> info) (1 second ago) dan@mccloy.info split off set_eeg_reference as its own mixin & add SetChannelsMixin to Info
c2fbb9f1a (4 minutes ago) dan@mccloy.info move setchannelsmixin to meas_info
```

which I just force-pushed here.
```

**Comment by drammock** [comment]
one final note: the optimal case involves *first creating 4-ish commits* for easy reviewing and then doing the squashing steps after the review.  There's no way (in this case at least) to have both a clean commit history (for later `git bisect` purposes) *and* a clean diff for commit-by-commit reviewing at the same time.

**Comment by drammock** [comment]
...aaand, I forgot the changelog 🤦🏻 stand by...

**Comment by drammock** [comment]
@larsoner assuming CIs come back green (they should 🤞🏻) this one is ready for a rebase-merge.

**Comment by larsoner** [comment]
Temporarily enabled rebase+merge, marked for auto-rebase+merge-when-green. We'll have to double check that the `ignore-revs` commit is still correct... in theory I think because your branch is up to date it shouldn't change the commit (the rebase should be a no-op) but we'll see!

**Comment by larsoner** [comment]
Failures are real in `mne/export`

**Comment by drammock** [comment]
> Failures are real in `mne/export`

WTF I checked that, why does it not show up in `git grep`?

```console
$ git grep _unit2human
mne/io/base.py:37:from .meas_info import _unit2human
mne/io/base.py:2311:            cols["unit"].append(_unit2human[ch["unit"]])
mne/io/meas_info.py:435:_unit2human = {
mne/io/meas_info.py:576:            if unit_old not in _unit2human:
mne/io/meas_info.py:582:                this_change = (_unit2human[unit_old], _unit2human[unit_new])
```

**Comment by drammock** [comment]
ahh, OK.  `pybv` is accessing our private API.

**Comment by alexrockhill** [comment]
Thanks for finishing the commit wrangling @drammock

**Pull request merged**
